pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
+
+        # 2. Assign the arguments of the constructor if needed.
+        if (hasattr(cls, "__init_attrs__")
+                and cls.__init_attrs__
+                and "__init__" in vars(cls)):
+            cls.__init__ = auto_assign(cls.__init__, sig=cls.__signature__)
+
+        # 3. Replace repr (if implemented) with its cached version.
+        if "__repr__" in vars(cls):
+            # If the class defines a custom `__repr__`, this will be the new
+            # `_repr` (which is cached). Decorator `cached_property` requires
+            # that `__set_name__` is called on it.
+            bmethod = functools.cached_property(cls.__repr__)
+            cls._repr = bmethod
+            bmethod.__set_name__(cls, "_repr")
+            # Fall back to using `__ccl_repr__` from `CCLObject`.
+            cls.__repr__ = cls.__ccl_repr__
+
+        # 4. Unlock instance on specific methods.
+        def Funlock(cls, name, mutate):
+            # Allow instance to change or mutate if method `name` is called.
+            func = vars(cls).get(name)
+            if func is not None:
+                newfunc = unlock_instance(mutate=mutate)(func)
+                setattr(cls, name, newfunc)
+
+        Funlock(cls, "__init__", False)
+        Funlock(cls, "update_parameters", True)
+
+    def __new__(cls, *args, **kwargs):
+        # Populate every instance with an `ObjectLock` as attribute.
+        instance = super().__new__(cls)
+        object.__setattr__(instance, "_object_lock", ObjectLock())
+        return instance
+
+    def __setattr__(self, name, value):
+        if self._object_lock.locked:
+            raise AttributeError("CCL objects can only be updated via "
+                                 "`update_parameters`, if implemented.")
+        object.__setattr__(self, name, value)
+
+    def update_parameters(self, **kwargs):
+        name = self.__class__.__qualname__
+        raise NotImplementedError(f"{name} objects are immutable.")
+
+    @functools.cached_property
+    def _repr(self):
+        # By default we use `__repr__` from `object`.
+        return object.__repr__(self)
+
+    @functools.cached_property
+    def _hash(self):
+        # `__hash__` makes use of the `repr` of the object,
+        # so we have to make sure that the `repr` is unique.
+        return hash(repr(self))
+
+    def __ccl_repr__(self):
+        # The custom `__repr__` is converted to a
+        # cached property and is replaced by this method.
+        return self._repr
+
+    __repr__ = __ccl_repr__
+
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, other):
+        # Two same-type objects are equal if their representations are equal.
+        if self.__class__ is not other.__class__:
+            return False
+        return repr(self) == repr(other)
+
+
+class CCLHalosObject(CCLObject):
+    """Base for halo objects. Automatically assign all ``__init__``
+    parameters as attributes.
```

Comment:
Again, it worries me that we're doing this.

Reply:
I have now introduced `__repr_attrs__` which acts as a class variable (hook) in place of the `auto_assign` decorator.
The idea is that instead of having to implement a tailor-made `__repr__` method for their class, developers can just specify which instance variables they need to be included in the `repr` via the `__repr_attrs__` hook, and the rest is done automatically.
As always, this can be overridden if they want to implement their own new shiny `repr`.

Reply:
For this to work, devs should make sure that the attribute names they pass in `__repr_attrs__` are defined as instance variables in `__init__`. However, there is no easy way to check down the whole MRO of a class to verify that all variables are indeed defined.

In order to not leave this hanging, I implemented a test to catch this kind of error during development, before it reaches the users. Before all the tests are run with `pytest`, I decorate every `__init__` which in subclasses of `CCLHalosObject` with a function that verifies that, once `__init__` has finished running, all variable names specified in `__repr_attrs__` exist. If they don't some tests will fail with a useful error. The test that will fail depends on where the erroneous class is used first.

Reply:
Note, that this check only happens in testing. The actual code doesn't have that. This check is developer-oriented, and it flags potential errors before they are merged to master.

Reply:
The test I'm talking about is in `tests/conftest.py` and uses the decorator I wrote in `tests/test_cclobject.py`.